### PR TITLE
New package variable: `opamfile`

### PR DESCRIPTION
**Pull request by rjbou**
superseeds #4387

### CODE REVIEW

File: src/state/opamPackageVar.ml [hunk 1]
```diff
@@ -305,6 +306,19 @@ let resolve st ?opam:opam_arg ?(local=OpamVariable.Map.empty) v =
...
+    | "opamfile-loc", Some opam ->
+      let repos_roots reponame =
+        match Hashtbl.find st.switch_repos.repos_tmp reponame with
+        | lazy repo_root -> repo_root
+        | exception Not_found ->
+          OpamRepositoryPath.root st.switch_global.root reponame
+      in
+      OpamFile.OPAM.get_metadata_dir ~repos_roots opam
+      |> OpamStd.Option.map (fun d ->
+          OpamFilename.Op.(d//"opam")
```
Comment (AltGr) [inline src/state/opamPackageVar.ml:318]: Hm, does this work for pinned packages ? Can't the file be named differently?

**Review by AltGr — COMMENTED** [review]
Thanks! Some concerns about the generality of implementation, and maybe we could find a better name ? (`-loc` doesn't sound straightforward to me... Why not just `opamfile` or `opam-file` ?)

File: src/state/opamPackageVar.ml (hunk 1 above)
Comment (rjbou) [reply src/state/opamPackageVar.ml:318]: for pinned packages, it's stored in the `overlay` dir, in the [file `opam`](https://github.com/ocaml/opam/blob/master/src/format/opamPath.ml#L254), and from repo, its the package dir with name `opam`.

**Comment by rjbou** [comment]
ok for me to rename it. `opamfile` can be misleading with the content of the opam file... first I thought about `opamfile-path`

**Comment by lefessan** [comment]
I think `opamfile` is ok, nobody would expect the content of the file as an argument to a hook script, no ?

File: src/state/opamPackageVar.ml (hunk 1 above)
Comment (AltGr) [reply src/state/opamPackageVar.ml:318]: aah right, it'll never get back to the original srcdir, we always have a mirror, got it! :+1:
maybe worth adding a comment ?

Comment (rjbou) [reply src/state/opamPackageVar.ml:318]: It's always there, in the current run of opam (repo dir are tmp ones), yes. In fact, it's the code of `OpamRepositoryState.get_root`, that is not reachable from `OpamPackageVar`.
+1 for the comment

**Pull request merged**